### ISSUE TITLE
Create ceterms:identifier

### ISSUE THREAD

**Issue by siuc-nate** [CredentialEngine/Schema-Development#666]
Based on our 4/1/2020 meeting:

We need a standalone property to point to https://credreg.net/ctdl/terms/IdentifierValue:

> **URI:** ceterms:identifier
> **Label:** Identifier
> **Definition:** Contextualized identifier for a resource.
> **Usage Note:** Where a specific property for an identifier exists, the best practice is to use that property instead of this one.
> **Domain:** ceterms:Credential (and subclasses), ceterms:LearningOpportunityProfile, ceterms:AssessmentProfile, ceterms:PathwayComponent subclasses, ceterms:TransferValueProfile (#683)
> **Range:** ceterms:IdentifierValue

We would also remove ceterms:codedNotation from the ceterms:PathwayComponent subclasses.

We will also need to explore the best way to clarify the distinction between ceterms:codedNotation and ceterms:identifierValueCode.

**Comment by mparsons-ce** [comment]
Credential, etc. currently have  the IdentifierValue property of: ceterms:versionIdentifier: https://credreg.net/ctdl/terms/versionIdentifier#versionIdentifier

Is this an additional IdentifierValue for credentials, etc?

**Comment by siuc-nate** [comment]
It's meant to cover cases where codedNotation isn't sufficient (i.e. there are multiple codes and/or codes that need to be distinguished by what they are), so, yes.

**Comment by siuc-nate** [comment]
@jeannekitchens @stuartasutton @mparsons-ce This was moved to the June milestone, probably during one of our meetings - but I don't remember what the reasoning was.  Does this still need to be held until June?  We want to use ceterms:identifier with transfer value profile.  

We could Add this to June as an unstable term just in case.  Would that work?

**Comment by siuc-nate** [comment]
Per our 5-19-2020 meeting:
This property seems potentially redundant with https://credreg.net/ctdl/terms/alternativeIdentifier - we should explore updating that property vs. creating a new one.

**Comment by stuartasutton** [comment]
How did it slip by me that codedNotation was included in PathwayComponent. It shouldn't be included. 

There is a clear distinction between an identifier and a codedNotation; although codedNotation might be considered a subproperty of identifier. The property codedNotation in CTDL-ASN and statementNotation in ASN are derived from the same notion as skos:notation (not to be confused with skos:note). 
>"A notation, **also known as classification code**, is a string of characters such as \"T58.5\" or \"303.4833\" used to uniquely identify a concept within the scope of a given concept scheme." 

While likely not treated as elegantly as SKOS suggests for notation, a codedNotation is decipherable by knowledgable humans familiar with the classification scheme--e.g., "CCSS.MATH.CONTENT.HSA.CED.A.2" from the Common Core State Standards and from above, T58.5 in the Library of Congress Classification System means T Subclass: Technology (General) 58.5 General works on Information systems and 303.4833 means Information technology--social effects in the Dewey Decimal Classification system. So, at a minimum, a codedNotation is decipherable (to the knowing) where vanilla identifiers are opaque.

**Comment by siuc-nate** [comment]
It was a side effect of rushing to get Pathways out as quickly as we did. We need to do more thorough testing, in particular with regard to encoding sample data using our proposed updates, before we push things out like that in the future.

At any rate, we will probably have to deal with the question of what to do with ceterms:codedNotation in any of the places it appears.  Do we have any reusable means of determining, for each such case, whether to:
1. Leave it as it is and do nothing
2. Add ceterms:identifier in parallel with ceterms:codedNotation in the same class
3. Remove ceterms:codedNotation and add ceterms:identifier
The follow-on for the last two being "What do we do with the existing data?" - a likely solution being to rewrite that data using ceterms:identifier with some arbitrary/boilerplate value for ceterms:identifierType.

**Comment by siuc-nate** [comment]
Per our 8/18/2020 meeting:

We will also update the definition for ceterms:codedNotation from
> Set of alpha-numeric symbols that uniquely identifies an item and supports its discovery and use.
to
> A notation, also known as classification code, is a string of characters such as "T58.5" or "303.4833" used to uniquely identify a concept within the scope of a given concept scheme.

**Comment by jeannekitchens** [comment]
@siuc-nate we need identifier for transfer value profile and for employment and training to jobs.  Is this the defintion we're moving foward with?  If not please update it with @stuartasutton @philbarker and this needs to be included with the domain models and terms proposals.   

What does "Contextualized" mean in the definition?  Also, I don't understand the usage note, doesn't make sense to me.

URI: ceterms:identifier
Label: Identifier
Definition: Contextualized identifier for a resource.
Usage Note: Where a specific property for an identifier exists, the best practice is to use that property instead of this one.
Domain: ceterms:Credential (and subclasses), ceterms:LearningOpportunityProfile, ceterms:AssessmentProfile, ceterms:PathwayComponent subclasses, ceterms:TransferValueProfile (#683)
Range: ceterms:IdentifierValue

**Comment by siuc-nate** [comment]
The definition is referencing the fact that ceterms:identifier points to a ceterms:IdentifierValue (class), which allows providing the text string of the identifier itself along with contextualizing information like name, type, and description.  It's not unlike a Concept or a DefinedTerm in that sense.

The comment basically means "If another property for this identifier exists, use that instead" (e.g. use ceterms:naics instead of an identifier that says "This is the NAICS code").

**Comment by philbarker** [comment]
Is this clearer?
**Definition**: Identifier for a resource and the named scheme for that identifier.
**Usage Note**: If a specific property exists for identifiers from a given scheme, use that property instead of this one.

**Comment by jgoodell2** [comment]
As someone viewing thread  midstream, the usage note is still confusing, but maybe with an example of a case where “specific property exists“ It might be clearer.

**Comment by stuartasutton** [comment]
...such as DUNS, FEIN, IPEDS ID, ISICV4, etc.

**Comment by siuc-nate** [comment]
While working on the NIST examples, which feature a variety of identifiers, I came up with the following logic for splitting these hairs. Consider the following to be food for thought rather than a formal proposal, as I am not yet 100% confident in it:

- A code is a `ceterms:codedNotation` if it is both described/defined by and exclusive to the object it is assigned to. 
  - Explanation: The object itself is what defines/describes the code.
  - Example: Even though `CCSS.MATH.CONTENT.HSA.CED.A.2` is part of a greater structure of codes, it is also exclusive to and defined/described by the very same standard it is attached to. There is no benefit to treating it as a reference to something external, hence a simple string `codedNotation` being sufficient. In other words, if you tried to look up the code, you would just find the same object you started with.

- A code is a `ceterms:identifier` if it is defined/described somewhere else, but exclusive to the object it is attached to.
  - Explanation: Some other source defines/describes the code, but the code only applies to the object the code is on.
  - Example 1: An Office of Personnel Management code in the NIST example is defined/described somewhere else (in other words, you could look up the code and find something other than just the object it's assigned to), but it applies exclusively to a single object (in the NIST example, a single/specific Work Role).
  - Example 2: An IPEDS ID, DUNS number, etc., is still exclusive to the object it's assigned to, but can be looked up and will contain information other than what would be contained on a schema:Organization that is referencing those codes.

- A code is an external category reference (called `ceterms:categoryReference` in the example at the bottom of this post) if it is defined/described somewhere else and is not exclusive to the object it is assigned to.
  - Explanation: The code is actually a way to look up a category of similar objects, usually called something that ends in "Type"
  - Example: Pretty much any property in CTDL that ends with "Type", including (and mostly inspired by) Industry Type, Occupation Type, and Instructional Program Type (which all reference a sort of "category" of Concepts/URIs)

What this looks like in practice (CredentialAlignmentObject omitted for brevity):
```
{
  "@id": "_:ConceptScheme_NISTGroupCode",
  "@type": "skos:ConceptScheme",
  "ceasn:name": { "en": "NIST Group Code" }
},
{
  "@id": "_:Concept_SP",
  "@type": "skos:Concept",
  "skos:prefLabel": { "en": "Securely Provision" },
  "skos:notation": "SP",
  "skos:inScheme": "_:ConceptScheme_NISTGroupCode"
},
{
  "@id": "_:Occupation_RSK",
  "@type": "ceterms:Occupation",
  "ceterms:name": { "en": "Risk Management" },
  "ceterms:description": { "en": "Oversees, evaluates, and supports the documentation, validation, assessment, and authorization processes necessary to assure that existing and new information technology (IT) systems meet the organization's cybersecurity and risk requirements. Ensures appropriate treatment of risk, compliance, and assurance from internal and external perspectives." },
  "ceterms:codedNotation": "RSK",
  "ceterms:categoryReference": [
    "_:Concept_SP"
  ],
  "ceterms:hasWorkRole": [
    "_:WorkRole_001"
  ]
},
{
  "@id": "_:WorkRole_001",
  "@type": "ceterms:WorkRole",
  "ceterms:name": { "en": "Authorizing Official/Designating Representative" },
  "ceterms:description": { "Senior official or executive with the authority to formally assume responsibility for operating an information system at an acceptable level of risk to organizational operations (including mission, functions, image, or reputation), organizational assets, individuals, other organizations, and the Nation (CNSSI 4009)." },
  "ceterms:codedNotation": "SP-RSK-001",
  "ceterms:identifier": [
    {
      "@type": "ceterms:IdentifierValue",
      "ceterms:name": { "en": "Office of Personnel Management Code" },
      "ceterms:description": { "en": "Fed Use" },
      "ceterms:identifierValueCode": "611"
    }
  ],
  "ceterms:hasTask": [
    "_:Task_T0145",
    "_:Task_T0221",
    "_:Task_T0371",
    "_:Task_T0495"
  ]
}
```
Note the use of "codedNotation", "identifier", and a general placeholder property to embody the third approach, "categoryReference" above.

The use of "categoryReference" here is an alternative to defining "Securely Provision" as a higher-level Occupation class, which technically also worked, but didn't seem to fit grammatically. It's also an alternative to using "occupationType", which didn't seem to fit either. Such a property might be useful in other cases, and/or applicable (if necessary) to Task.

I'll keep experimenting with this approach as I work through the other occupation-related examples, and may end up changing my mind if I run into issues.

**Comment by siuc-nate** [comment]
I also noticed that a `ceterms:IdentifierValue` is _painfully_ close to a `skos:Concept`. Both have a preferred label, a definition, a notation code, and are usually part of some greater structure/scheme/framework.  The only thing that differentiates them is that an IdentifierValue's code is (at least in the examples so far) unique to a single object, whereas a Concept can be broadly applicable. But I'm not sure if that's a sufficiently strong argument for not using the skos:Concept class anyway.

**Comment by siuc-nate** [comment]
There are two special cases that show up in the Navy example:

1. Composite identifiers - These are where a sequence of codes is itself a code. In the Navy example, `HGONGAA` expands out to 6 different references to 6 different concepts from 6 different schemes (`H`,`G`,`O`,`N`,`G`,`AA`) that contain various detailed information about the task.  The composite code is not itself does not reference a "real" entity; it is a string constructed from several other codes. It is also not unique to the object, as other objects could be assigned the same combination of separate codes and thus end up with the same composite identifier. So it sits awkwardly between `ceterms:identifier` and `ceterms:categoryReference`. 

2. Partially-defined code lists - There are a few parts of the GEIA0007 schema that use partially-defined codes (i.e. a two or three character code where a few combinations have a predetermined meaning but other combinations are intentionally left undefined so that they can be determined dynamically as needed).  This is a really rare edge case that can probably just be handled with `ceterms:identifier` but it may be useful to be able to reference a `skos:Concept` for those codes that _are_ predefined.

**Comment by philbarker** [comment]
> I also noticed that a ceterms:IdentifierValue is painfully close to a skos:Concept. 

Not really. "A SKOS concept can be viewed as an idea or notion; a unit of thought." https://www.w3.org/TR/skos-reference/#concepts An IdentifierValue is a way of identifying things. 

SKOS Concepts need identifying, and so they have the components of an IdentifierValue.

It's maybe worth noting the difference between identifying something and classifying it. The ISBN for a book identifies the edition; the Dewey Decimal Code classifies the book. The Dewey Decimal Code also identifies a Concept in the DDC scheme, but it absolutely does not _identify_ a book. So I wouldn't get too hung up on whether something is an identifier without asking "of what?"

**Comment by stuartasutton** [comment]
+1 to https://github.com/CredentialEngine/Schema-Development/issues/666#issuecomment-681839710

**Comment by siuc-nate** [comment]
See also #599

**Comment by siuc-nate** [comment]
This issue has been superseded by #715. Go there for the resolution.
